File: fastapi/src/routes/project_routes.py
from fastapi import APIRouter, Depends, Response

from src.schemas.common_schemas import ErrorResponse
from src.schemas.project_schemas import (
    ProjectCreate,
    ProjectListQuery,
    ProjectListResponse,
    ProjectResponse,
    ProjectUpdate,
)
from src.services.project_service import ProjectService, get_project_service

router = APIRouter(prefix="/projects", tags=["Projects"])


@router.get(
    "",
    response_model=ProjectListResponse,
    summary="List projects",
)
async def list_projects(
    query: ProjectListQuery = Depends(),
    service: ProjectService = Depends(get_project_service),
) -> ProjectListResponse:
    items, meta = await service.list(
        search=query.search,
        status=query.status,
        limit=query.limit,
        offset=query.offset,
    )
    return ProjectListResponse(
        data=[
            ProjectResponse(
                id=p.id,
                name=p.name,
                description=p.description,
                status=p.status,
                owner_id=p.owner_id,
                created_at=p.created_at,
                updated_at=p.updated_at,
            )
            for p in items
        ],
        meta=meta,
    )


@router.get(
    "/{id}",
    response_model=ProjectResponse,
    summary="Get project by ID",
    responses={404: {"model": ErrorResponse}},
)
async def get_project(
    id: str,
    service: ProjectService = Depends(get_project_service),
) -> ProjectResponse:
    from uuid import UUID

    project = await service.find_by_id(UUID(id))
    return ProjectResponse(
        id=project.id,
        name=project.name,
        description=project.description,
        status=project.status,
        owner_id=project.owner_id,
        created_at=project.created_at,
        updated_at=project.updated_at,
    )


@router.post(
    "",
    response_model=ProjectResponse,
    status_code=201,
    summary="Create a project",
)
async def create_project(
    params: ProjectCreate,
    service: ProjectService = Depends(get_project_service),
) -> ProjectResponse:
    project = await service.create(params)
    return ProjectResponse(
        id=project.id,
        name=project.name,
        description=project.description,
        status=project.status,
        owner_id=project.owner_id,
        created_at=project.created_at,
        updated_at=project.updated_at,
    )


@router.patch(
    "/{id}",
    response_model=ProjectResponse,
    summary="Update a project",
    responses={404: {"model": ErrorResponse}},
)
async def update_project(
    id: str,
    params: ProjectUpdate,
    service: ProjectService = Depends(get_project_service),
) -> ProjectResponse:
    from uuid import UUID

    project = await service.update(UUID(id), params)
    return ProjectResponse(
        id=project.id,
        name=project.name,
        description=project.description,
        status=project.status,
        owner_id=project.owner_id,
        created_at=project.created_at,
        updated_at=project.updated_at,
    )


@router.delete(
    "/{id}",
    status_code=204,
    summary="Delete a project",
    responses={404: {"model": ErrorResponse}},
)
async def delete_project(
    id: str,
    service: ProjectService = Depends(get_project_service),
) -> Response:
    from uuid import UUID

    await service.delete(UUID(id))
    return Response(status_code=204)

File: hono-bun/src/routes/project-routes.ts
import { OpenAPIHono, createRoute, z } from "@hono/zod-openapi";

import type { HonoEnv } from "@/core/hono-env";
import { ErrorSchema } from "@/schemas/common-schemas";
import {
  ProjectCreateSchema,
  ProjectListQuerySchema,
  ProjectListResponseSchema,
  ProjectSchema,
  ProjectUpdateSchema,
} from "@/schemas/project-schemas";
import { projectService } from "@/services/project-service";

const projectRoutes = new OpenAPIHono<HonoEnv>();

const listProjects = createRoute({
  method: "get",
  path: "/projects",
  tags: ["Projects"],
  summary: "List projects",
  request: { query: ProjectListQuerySchema },
  responses: {
    200: {
      description: "List of projects",
      content: { "application/json": { schema: ProjectListResponseSchema } },
    },
  },
});

projectRoutes.openapi(listProjects, async (c) => {
  const query = c.req.valid("query");
  const { items, meta } = await projectService.list({
    search: query.search,
    status: query.status,
    limit: query.limit,
    offset: query.offset,
  });
  return c.json(
    {
      data: items.map((p) => ({
        id: p.id,
        name: p.name,
        description: p.description,
        status: p.status,
        ownerId: p.ownerId,
        createdAt: p.createdAt.toISOString(),
        updatedAt: p.updatedAt.toISOString(),
      })),
      meta,
    },
    200,
  );
});

const getProject = createRoute({
  method: "get",
  path: "/projects/{id}",
  tags: ["Projects"],
  summary: "Get project by ID",
  request: { params: z.object({ id: z.string().uuid() }) },
  responses: {
    200: {
      description: "Project detail",
      content: { "application/json": { schema: ProjectSchema } },
    },
    404: {
      description: "Not found",
      content: { "application/json": { schema: ErrorSchema } },
    },
  },
});

projectRoutes.openapi(getProject, async (c) => {
  const { id } = c.req.valid("param");
  const p = await projectService.findById(id);
  return c.json(
    {
      id: p.id,
      name: p.name,
      description: p.description,
      status: p.status,
      ownerId: p.ownerId,
      createdAt: p.createdAt.toISOString(),
      updatedAt: p.updatedAt.toISOString(),
    },
    200,
  );
});

const createProject = createRoute({
  method: "post",
  path: "/projects",
  tags: ["Projects"],
  summary: "Create a project",
  request: {
    body: {
      content: { "application/json": { schema: ProjectCreateSchema } },
    },
  },
  responses: {
    201: {
      description: "Created",
      content: { "application/json": { schema: ProjectSchema } },
    },
  },
});

projectRoutes.openapi(createProject, async (c) => {
  const body = c.req.valid("json");
  const p = await projectService.create({
    name: body.name,
    description: body.description,
    status: body.status,
    ownerId: c.get("user")?.id,
  });
  return c.json(
    {
      id: p.id,
      name: p.name,
      description: p.description,
      status: p.status,
      ownerId: p.ownerId,
      createdAt: p.createdAt.toISOString(),
      updatedAt: p.updatedAt.toISOString(),
    },
    201,
  );
});

const updateProject = createRoute({
  method: "patch",
  path: "/projects/{id}",
  tags: ["Projects"],
  summary: "Update a project",
  request: {
    params: z.object({ id: z.string().uuid() }),
    body: {
      content: { "application/json": { schema: ProjectUpdateSchema } },
    },
  },
  responses: {
    200: {
      description: "Updated",
      content: { "application/json": { schema: ProjectSchema } },
    },
    404: {
      description: "Not found",
      content: { "application/json": { schema: ErrorSchema } },
    },
  },
});

projectRoutes.openapi(updateProject, async (c) => {
  const { id } = c.req.valid("param");
  const body = c.req.valid("json");
  const p = await projectService.update(id, body);
  return c.json(
    {
      id: p.id,
      name: p.name,
      description: p.description,
      status: p.status,
      ownerId: p.ownerId,
      createdAt: p.createdAt.toISOString(),
      updatedAt: p.updatedAt.toISOString(),
    },
    200,
  );
});

const deleteProject = createRoute({
  method: "delete",
  path: "/projects/{id}",
  tags: ["Projects"],
  summary: "Delete a project",
  request: { params: z.object({ id: z.string().uuid() }) },
  responses: {
    204: { description: "Deleted" },
    404: {
      description: "Not found",
      content: { "application/json": { schema: ErrorSchema } },
    },
  },
});

projectRoutes.openapi(deleteProject, async (c) => {
  const { id } = c.req.valid("param");
  await projectService.delete(id);
  return new Response(null, { status: 204 });
});

export default projectRoutes;
